### ISSUE TITLE
Which dataset was used?

### ISSUE THREAD

**Issue by ma7555** [ozora-ogino/wav2spk#1]
Hello Ozora,

I am trying to implement the same, I was wondering which dataset you used during training?

Also, I noted that this implentation is different from the paper's implementation by the following

1. 'SGD' optimizer is used instead of 'Adam'
2. Additive Margin Softmax loss function instead of categorical crossentropy.
3. You are using 30916 samples per input, with sampling rate 16000, which corresponds to ~1930ms, while the original paper mentioned using 400ms segments.

Please note that I have found a TF1 implentation of AM-Softmax here https://www.programmersought.com/article/19835701490/ and tried adapting it for TF2 but it throws an error that I am still debugging.

Results are very weird for me on my own dataset and wanted to try on the one you used.

Thanks

**Comment by ozora-ogino** [comment]
Hi ma,

Thank you for contacting me.

What you said is right.
My implementation is not same as original.
However I used voxceleb1 and voxceleb2 for training, which are same as original paper.

I guess [TF1 implentation of AM-Softmax](https://www.programmersought.com/article/19835701490/) isn't compatible with TensorFlow 2.0. So I'm trying to adapting it for TensorFlow2.0 too.

Our purpose and problem we faced is same, so we can work together if you can.
If you like to collaborate with me, share your implementation.

Thank you.

**Comment by ma7555** [comment]
Hi @ozora-ogino,

So, I have tried with the voxceleb1 dataset now, the model doesn't converge at all, loss is typically the same and val_acc is 0.
I am wondering you managed to achieve any results on this model?

```
Epoch 1/200
1046/1046 [==============================] - 201s 193ms/step - loss: 6.8711 - accuracy: 0.0071 - val_loss: 11.5359 - val_accuracy: 0.0000e+00
Epoch 2/200
1046/1046 [==============================] - 184s 175ms/step - loss: 6.8499 - accuracy: 0.0075 - val_loss: 11.8020 - val_accuracy: 0.0000e+00
Epoch 3/200
1046/1046 [==============================] - 183s 175ms/step - loss: 6.8472 - accuracy: 0.0075 - val_loss: 12.6955 - val_accuracy: 0.0000e+00
Epoch 4/200
1046/1046 [==============================] - 183s 175ms/step - loss: 6.8457 - accuracy: 0.0075 - val_loss: 12.8120 - val_accuracy: 0.0000e+00
```

Regarding the AM-Softmax, i am still debugging. this is a slighly modified version that I am working on

```
class AMSoftmax(Layer):
    def __init__(self, units, s, m,
                 kernel_initializer='glorot_uniform'
                 ):

        super(AMSoftmax, self).__init__()
        self.units = units
        self.s = s
        self.m = m
        self.kernel_initializer = initializers.get(kernel_initializer)

    def build(self, input_shape):
        super().build(input_shape[0])
        assert len(input_shape) >= 2
        input_dim = input_shape[-1]

        self.kernel = self.add_weight(shape=(input_dim, self.units),
                                      initializer=self.kernel_initializer,
                                      name='kernel')

    def call(self, inputs):

        inputs = tf.math.l2_normalize(inputs, axis=-1)
        self.kernel = tf.math.l2_normalize(self.kernel, axis=(0, 1))   # W Normalization

        dis_cosin = tf.linalg.matmul(inputs, self.kernel)
        psi = dis_cosin - self.m

        e_costheta = tf.math.exp(self.s * dis_cosin)
        e_psi = tf.math.exp(self.s * psi)
        sum_x = tf.math.reduce_sum(e_costheta, axis=-1, keepdims=True)

        temp = e_psi - e_costheta
        temp = temp + sum_x

        output = e_psi / temp
        return output
```
if you comment this line, things will work, I will continue debugging and let you know the results.
```
        self.kernel = tf.math.l2_normalize(self.kernel, axis=(0, 1))   # W Normalization
```

Also, the pooling layer in the paper is different and there is no implemntation for it in TF, I didn't read anything about it yet.

I would like to collaborate so please share what do you think but let me know what happens when you train on this exact model with voxceleb1

Thanks!

**Comment by ozora-ogino** [comment]
Hi @ma7555 

I guess my implementation is wrong.
I haven't got the accuracy such as the results of original paper yet.

We need to implement wav2spk model correctly.
I published new issue #2 which lists what we need to implement. Please check it.

I believe AM-Softmax is not a serious problem because a big issue we're facing now  is that training losses don't decrease.
And that may be caused by wrong implementation of temporal gating or stat pooling.
So let's try to implement them fast.

BTW, if you have your own implementation of Wav2Spk, let me see please.

Thank you and stay safe!
